### PR TITLE
PWA-2113: [GraphQL] Create mutation to set billing address on negotiable quote

### PR DESCRIPTION
**Pull request by cspruiell**
## Problem

We cannot currently support end-to-end checkout of negotiable quotes in GraphQL because several mutations are missing, including one to set the billing address on the negotiable quote. (Other missing mutations will be addressed separately.)

## Solution

Add the setNegotiableQuoteBillingAddress mutation to the NegotiableQuoteGraphQl schema and return the full negotiable quote as the output.

### CODE REVIEW

File: design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls [hunk 1]
```diff
@@ -359,3 +371,7 @@ type NegotiableQuoteUser @doc(description: "A limited view of a Buyer or Seller
 type StoreConfig {
     is_negotiable_quote_active: Boolean @doc(description: "Indicates if negotiable quote functionality is enabled.")
 }
+
+type SetNegotiableQuoteBillingAddressOutput {
+    quote: NegotiableQuote
```
Comment (dthampy) [inline design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls:376]: I believe we should add billing_address field  in the NegotiableQuote object if we are only returning quote to match with the proposal in [this PR](https://github.com/magento/architecture/pull/495/files) or add it as an additional output field in the SetNegotiableQuoteBillingAddressOutput.
`billing_address: BillingCartAddress`

File: design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls [hunk 2]
```diff
@@ -81,6 +85,14 @@ input SetNegotiableQuoteShippingAddressInput {
     )
 }
 
+input SetNegotiableQuoteBillingAddressInput {
+    quote_uid: ID! @doc(description: "ID obtained from NegotiableQuote type")
+    customer_address_id: Int @doc(
+        description: "ID obtained from the CustomerAddress type. A new address can be added using Mutation.createCustomerAddress"
+    )
+    new_customer_address: CustomerAddressInput @doc(description: "New address to be added")
```
Comment (dthampy) [inline design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls:93]: `SetBillingAddressonCartInput` takes in the CartAddressInput for billing_address. Do we need all the additional fields provided by `CustomerAddressInput` for negotiableQuote, new_customer_address?
@pdohogne-magento  @amolina-magento : what do you guys think?

File: design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls [hunk 3]
```diff
@@ -81,6 +85,14 @@ input SetNegotiableQuoteShippingAddressInput {
...
+}
```
Comment (dthampy) [inline design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls:94]: Also, looks like, we are missing the `same_as_shipping` field as well

File: design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls (hunk 1 above)
Comment (cspruiell) [reply design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls:374]: added

File: design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls (hunk 2 above)
Comment (cspruiell) [reply design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls]: I've changed these inputs to match Cart, which also address the `same_as_shipping` comment below

Comment (amolina-magento) [reply design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls]: The `SetNegotiableQuoteBillingAddressInput` has been updated to reuse the `BillingAddressInput` from QuoteGraphQL. Given our recent conversations, this looks good for now.

**Review by dthampy — APPROVED** [review]

File: design-documents/graph-ql/coverage/cart/CartAddressOperations.graphqls [hunk 4]
```diff
@@ -22,7 +22,8 @@ input SetBillingAddressOnCartInput {
 input BillingAddressInput {
     customer_address_id: Int
     address: CartAddressInput
-    use_for_shipping: Boolean
+    use_for_shipping: Boolean @doc(description: "Indicates whether to additionally set the shipping address based on the provided billing address")
+    same_as_shipping: Boolean @doc(description: "Indicates whether to set the billing address based on the existing shipping address on the cart")
 }
```
Comment (cspruiell) [inline design-documents/graph-ql/coverage/cart/CartAddressOperations.graphqls:27]: this is updated to bring it in sync with the current schema

File: design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls [hunk 5]
```diff
@@ -81,6 +85,32 @@ input SetNegotiableQuoteShippingAddressInput {
...
+    quote_uid: ID! @doc(description: "ID obtained from NegotiableQuote type")
+    billing_address: NegotiableQuoteBillingAddressInput!
+}
+
+input NegotiableQuoteBillingAddressInput {
+    customer_address_id: Int
+    address: NegotiableQuoteAddressInput
+    use_for_shipping: Boolean @doc(description: "Indicates whether to additionally set the shipping address based on the provided billing address")
+    same_as_shipping: Boolean @doc(description: "Indicates whether to set the billing address based on the existing shipping address on the cart")
+}
```
Comment (dthampy) [inline design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls:98]: change Cart to negotiable quote

File: design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls [hunk 6]
```diff
@@ -185,6 +215,7 @@ type NegotiableQuote {
     # attachments: [AttachmentContent]
     comments: [NegotiableQuoteComment!]
     history: [NegotiableQuoteHistoryEntry!]
+    billing_address: BillingCartAddress
```
Comment (pdohogne-magento) [inline design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls:218]: Need Negotiable Quote-specific typing.  Collaborate with @amolina-magento for the upstream address type.

Comment (amolina-magento) [inline design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls:218]: We need a Negotiable Quote specific version of this
`NegotiableQuoteAddressInterface` -> `NegotiableQuoteBillingAddress` -> `NegotiableQuoteAddressRegion`/`NegotiableQuoteAddressCountry`

File: design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls [hunk 7]
```diff
@@ -185,13 +215,39 @@ type NegotiableQuote {
...
+    firstname: String
+    lastname: String
+    company: String
+    street: [String]
+    city: String
+    region: NegotiableQuoteAddressRegion
+    postcode: String
+    country: NegotiableQuoteAddressCountry
+    telephone: String
+}
```
Comment (pdohogne-magento) [inline design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls:239]:
```suggestion
interface NegotiableQuoteAddressInterface {
    firstname: String!
    lastname: String!
    company: String
    street: [String!]!
    city: String!
    region: NegotiableQuoteAddressRegion
    postcode: String
    country: NegotiableQuoteAddressCountry!
    telephone: String
}
```

File: design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls [hunk 8]
```diff
@@ -185,13 +215,39 @@ type NegotiableQuote {
...
+
+type NegotiableQuoteAddressRegion {
+    code: String
+    label: String
+}
```
Comment (pdohogne-magento) [inline design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls:244]:
```suggestion
type NegotiableQuoteAddressRegion {
    code: String
    label: String
    region_id: Int
}
```

File: design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls [hunk 9]
```diff
@@ -185,13 +215,39 @@ type NegotiableQuote {
...
+
+type NegotiableQuoteAddressCountry {
+    code: String
+    label: String
```
Comment (pdohogne-magento) [inline design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls:248]:
```suggestion
    code: String!
    label: String!
```

File: design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls (hunk 1 above)
Comment (pdohogne-magento) [reply design-documents/graph-ql/coverage/b2b/negotiableQuotes.graphqls:420]:
```suggestion
    quote: NegotiableQuote!
```

**Review by amolina-magento — APPROVED** [review]

**Review by pdohogne-magento — APPROVED** [review]

**Comment by pdohogne-magento** [comment]
Changes included in https://github.com/magento/architecture/commit/22de863cea586418ab468fb6970914d20dc38676

**Pull request closed without merging**